## Downloading new data and adding season + physical variables to dataframe so new csv can be read in and ready to go in every notebook

In [1]:
import numpy as np
from oneargopy.OneArgo import Argo
from argopy import DataFetcher as ArgoDataFetcher
import pandas as pd
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import matplotlib.pyplot as plt
import cartopy.feature as cfeature
import cmocean
import matplotlib.colors as mcolors
import matplotlib.dates as mdates
import matplotlib.cm as cm # new package for the colorbar
import matplotlib.patches as mpatches
from cartopy.mpl.ticker import LongitudeFormatter, LatitudeFormatter
from matplotlib.ticker import LogLocator, FormatStrFormatter
from matplotlib.colors import LogNorm
import gsw

In [2]:
argo = Argo('argo_config.json')

Starting initialize process...
Your current download settings are: 
[DownloadSettings] -> 
Base Directory: ., 
Subdirectories: ['Index', 'Meta', 'Tech', 'Traj', 'Profiles'], 
Index Files: ['ar_index_global_traj.txt', 'ar_index_global_tech.txt', 'ar_index_global_meta.txt', 'ar_index_global_prof.txt', 'argo_synthetic-profile_index.txt'], 
Verbose Setting: True, 
Max Attempts: 5, 
Keep Index In Memory: True, 
Float Type: bgc

Your current source settings are: 
[Source Settings] -> 
Hosts: ['https://data-argo.ifremer.fr/', 'https://usgodae.org/ftp/outgoing/argo/'], 
Available Variables: None, 
Data Assimilation Centers: None

Checking for subdirectories...
The Index directory already exists
The Meta directory already exists
The Tech directory already exists
The Traj directory already exists
The Profiles directory already exists

ar_index_global_traj.txt does not need to be updated yet.
ar_index_global_tech.txt does not need to be updated yet.
ar_index_global_meta.txt does not need to be up

In [3]:
pressure = argo.select_profiles(lon_lim=[150,180], lat_lim=[-68,-58], type = 'bgc', sensor = 'CHLA') # all bgc floats within region

Starting select_profiles...
Validating parameters...
Validating longitude and latitude limits...
Validating start and end dates...
Validating 'type' keyword argument...
Preparing float data for filtering...
Filtering through 20367 floats
There are 3351908 profiles associated with these floats

Sorting floats for those within the geographic range...
Creating point list from profiles...
Creating polygon...
79 floats fall within the geographic range
2605 profiles associated with those floats
Applying outside=None constraints...
79 floats selected
2605 profiles selected according to time and space constraints
Floats Selected: dict_keys([1901135, 1901153, 1901155, 1901157, 1902662, 2901936, 2903454, 3902129, 4903658, 5900345, 5901178, 5901645, 5901648, 5901698, 5901699, 5903242, 5903248, 5903260, 5903630, 5903717, 5903719, 5903720, 5903721, 5903722, 5903723, 5903939, 5904104, 5904105, 5904179, 5904481, 5904482, 5904483, 5904598, 5904599, 5904676, 5904677, 5904682, 5904688, 5905096, 5905097,

## New way of downloading data to avoid masked constant error

problem floats: 3902129, 6902814

Dataframe ran fine without the two floats above: they gave an "unsupported type for timedelta days component: MaskedConstant" error

In [4]:
# float ids from argo.select_profiles function for extended AAR region
float_ids = [1901135, 1901153, 1901155, 1901157, 1902662, 2901936, 2903454, 4903658, 5900345, 5901178, 5901645, 5901648, 5901698, 5901699, 5903242, 5903248, 5903260, 5903630, 5903717, 5903719, 5903720, 5903721, 5903722, 5903723, 5903939, 5904104, 5904105, 5904179, 5904481, 5904482, 5904483, 5904598, 5904599, 5904676, 5904677, 5904682, 5904688, 5905096, 5905097, 5905098, 5905099, 5905100, 5905101, 5905102, 5905103, 5905232, 5905372, 5905373, 5905374, 5905375, 5905376, 5905377, 5905379, 5905396, 5905555, 5905556, 5905557, 5905638, 5906207, 5906221, 5906442, 5906509, 5906581, 5906582, 5906583, 5906624, 6901004, 6902885, 6902891, 6903190, 6903233, 6903768, 6903777, 6990587, 7902154, 7902155, 7902379]


In [6]:
df = argo.load_float_data(float_ids, variables=['DOXY','TEMP','PSAL','CHLA', 'BBP700','NITRATE','PH_IN_SITU_TOTAL'])

Validating passed floats...
Validating passed 'variables'...
1901135_Sprof.nc does not need to be updated yet.
1901153_Sprof.nc does not need to be updated yet.
1901155_Sprof.nc does not need to be updated yet.
1901157_Sprof.nc does not need to be updated yet.
1902662_Sprof.nc does not need to be updated yet.
2901936_Sprof.nc does not need to be updated yet.
2903454_Sprof.nc does not need to be updated yet.
4903658_Sprof.nc does not need to be updated yet.
5900345_Sprof.nc does not need to be updated yet.
5901178_Sprof.nc does not need to be updated yet.
5901645_Sprof.nc does not need to be updated yet.
5901648_Sprof.nc does not need to be updated yet.
5901698_Sprof.nc does not need to be updated yet.
5901699_Sprof.nc does not need to be updated yet.
5903242_Sprof.nc does not need to be updated yet.
5903248_Sprof.nc does not need to be updated yet.
5903260_Sprof.nc does not need to be updated yet.
5903630_Sprof.nc does not need to be updated yet.
5903717_Sprof.nc does not need to be up

In [7]:
df.to_csv('/Users/lilah/Documents/IBIS_Project/Data_Summer/new_unedited_bgc_profiles_BR_extended_area_all_time.csv', index = 'False')
# df.to_csv('/Users/steviewalker/Documents/IBIS_Project/Data_Summer/new_bgc_profiles_BR_extended_area_all_time.csv', index = 'False')

In [8]:
df = pd.read_csv('/Users/lilah/Documents/IBIS_Project/Data_Summer/new_unedited_bgc_profiles_BR_extended_area_all_time.csv')

/var/folders/sr/nn_fjjln6s33md98tyfc3rd40000gn/T/ipykernel_42674/3263918958.py:1: DtypeWarning: Columns (0: LATITUDE, 1: LONGITUDE) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/Users/lilah/Documents/IBIS_Project/Data_Summer/new_unedited_bgc_profiles_BR_extended_area_all_time.csv')


## Convert to datetime, add month, year, season, MLP and physical variable columns (absolute salinity, conservative temperaure, potential density, in situ density, and buoyancy frequency

Might redo how MLP is calculated based on Terrats et al 2023 paper?

Datetime and latitude/longitude NaN values

In [9]:
# NEW DATA

#convert to datetime
df['DATE'] = pd.to_datetime(df['DATE'],format = 'mixed',utc=True)
df = df.dropna(subset=["DATE"]) #drop any cycles with no date or lat/lon 

#fix lat lon values with '--' (replace with nan) 
df['LATITUDE'] = pd.to_numeric(df['LATITUDE'], errors='coerce')
df['LONGITUDE'] = pd.to_numeric(df['LONGITUDE'], errors='coerce')


Month, year, and season column added

In [10]:
#create season columns
df["year"] = df["DATE"].dt.year.astype("Int64")
df["month"] = df["DATE"].dt.month.astype("Int64")

def assign_season(row):
    m = int(row["month"])
    y = int(row["year"])
    if m >= 9:
        return f"{y}-{y+1}"
    elif m <= 4:
        return f"{y-1}-{y}"
    else:
        return f"{y}"

df["season"] = df.apply(assign_season, axis=1)

Physical variable columns added (SA, CT, potential density, in-situ density, buoyancy frequency)

In [22]:
# define function for calculating SA, CT, potential dens, in-situ dens, buoyancy frequency
def calc_phys_vars(df):
        
    # Columns needed
    cols = ['PSAL_ADJUSTED','TEMP_ADJUSTED' ,'PRES_ADJUSTED', 'LATITUDE', 'LONGITUDE']
    
    # only do calculation on good qc T and S data - create nan values where not good data
    qc_flags = [1, 2,5, 8]
    mask_qc = df['PSAL_ADJUSTED_QC'].isin(qc_flags) & df['TEMP_ADJUSTED_QC'].isin(qc_flags)
    df['PSAL_ADJUSTED'] = df['PSAL_ADJUSTED'].where(mask_qc)
    df['TEMP_ADJUSTED'] = df['TEMP_ADJUSTED'].where(mask_qc)    
    
    # mask valid (non-null) rows
    mask = df[cols].notnull().all(axis=1)
    
    # initialize outputs as NaN
    SA = np.full(len(df), np.nan)
    CT = np.full(len(df), np.nan)
    rho = np.full(len(df), np.nan)
    sigma0 = np.full(len(df), np.nan)
    nsq = np.full(len(df), np.nan)
    
    # extract numeric arrays for valid rows
    SP = df.loc[mask, 'PSAL_ADJUSTED'].astype(float).to_numpy()
    PT = df.loc[mask, 'TEMP_ADJUSTED'].astype(float).to_numpy()
    pres = df.loc[mask, 'PRES_ADJUSTED'].astype(float).to_numpy()
    LAT = df.loc[mask, 'LATITUDE'].astype(float).to_numpy()
    LON = df.loc[mask, 'LONGITUDE'].astype(float).to_numpy()
    
    # calculate variables using gsw
    SA_valid = gsw.SA_from_SP(SP=SP, p=pres, lon=LON, lat=LAT)
    CT_valid = gsw.CT_from_t(SA=SA_valid, t=PT, p=pres)
    rho_valid = gsw.rho(SA=SA_valid, CT=CT_valid, p=pres)
    sigma0_valid = gsw.sigma0(SA=SA_valid, CT=CT_valid)
    nsq_valid = gsw.Nsquared(SA=SA_valid, CT=CT_valid, p=pres,lat=LAT)
    
    # store back into full-length arrays
    SA[mask] = SA_valid
    CT[mask] = CT_valid
    rho[mask] = rho_valid
    sigma0[mask] = sigma0_valid
    np.putmask(nsq, mask, nsq_valid) # found on stack overflow because nsq was 2 dimensional so previous method wasn't working
    
    # sdd results to the dataframe
    df = df.copy()
    df['SA'] = SA
    df['CT'] = CT
    df['rho'] = rho
    df['sigma0'] = sigma0
    df['nsq'] = nsq
    
    return df

df_edited = calc_phys_vars(df)

/opt/anaconda3/envs/IBIS_Project/lib/python3.13/site-packages/gsw/stability.py:86: RuntimeWarning: divide by zero encountered in divide
  N2 = ((g_local**2) / (specvol_mid * db_to_pa * dp))


MLP column added - eventually change how it's calculated?

In [19]:
def calc_mlp_density(df_profile, delta_sigma=0.03):
    #calc MLD for a single profile using density threshold
    df_profile = df_profile.sort_values('PRES_ADJUSTED')

    # get SA, CT, and potential density values
    SA = df_profile['SA']
    CT = df_profile['CT']
    sigma0 = df_profile['sigma0']

    sigma_surface = sigma0.iloc[0]
    d_sigma = sigma0 - sigma_surface

    # find first depth where density difference exceeds threshold
    idx = np.argmax(d_sigma > delta_sigma)
    if d_sigma.iloc[idx] <= delta_sigma:
        return np.nan  # no threshold reached
    return df_profile['PRES_ADJUSTED'].iloc[idx]

mld = (df_edited.groupby(['WMOID', 'CYCLE_NUMBER']).apply(calc_mlp_density).reset_index(name='MLP'))
df_full_edited = df_edited.merge(mld, on=['WMOID', 'CYCLE_NUMBER'], how='left')

Save the dataframe with new columns added so you can just read in the csv in each notebook

In [20]:
df_full_edited.to_csv('/Users/lilah/Documents/IBIS_Project/Data_Summer/new_bgc_profiles_BR_extended_area_all_time.csv', index = 'False')

In [21]:
df_full_edited

,Unnamed: 0,WMOID,CYCLE_NUMBER,DIRECTION,DATE,DATE_QC,LATITUDE,LONGITUDE,POSITION_QC,PRES,...,PH_IN_SITU_TOTAL_ADJUSTED_ERROR,year,month,season,SA,CT,rho,sigma0,nsq,MLP
0,0,1901135,1,A,2008-11-26 23:49:02.000028+00:00,1,-43.9840,72.3760,1,6.20,...,NaN,2008,11,2008-2009,34.254127,9.551515,1026.351057,26.322117,1.487167e-06,100.60
1,1,1901135,1,A,2008-11-26 23:49:02.000028+00:00,1,-43.9840,72.3760,1,10.50,...,NaN,2008,11,2008-2009,34.255093,9.552023,1026.371163,26.322783,-8.333599e-07,100.60
2,2,1901135,1,A,2008-11-26 23:49:02.000028+00:00,1,-43.9840,72.3760,1,15.10,...,NaN,2008,11,2008-2009,34.255108,9.554516,1026.391560,26.322385,-1.304368e-07,100.60
3,3,1901135,1,A,2008-11-26 23:49:02.000028+00:00,1,-43.9840,72.3760,1,20.10,...,NaN,2008,11,2008-2009,34.255115,9.554964,1026.414093,26.322317,-6.799600e-07,100.60
4,4,1901135,1,A,2008-11-26 23:49:02.000028+00:00,1,-43.9840,72.3760,1,25.40,...,NaN,2008,11,2008-2009,34.254085,9.552394,1026.437673,26.321941,1.803892e-07,100.60
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6484485,6484485,7902379,16,A,2026-07-21 13:09:29.001408+00:00,1,-56.5413,161.4417,1,1799.14,...,1.336424,2026,7,2026,34.905167,1.316037,1036.150275,27.816317,7.519300e+02,172.32
6484486,6484486,7902379,16,A,2026-07-21 13:09:29.001408+00:00,1,-56.5413,161.4417,1,1849.45,...,1.337056,2026,7,2026,34.905236,1.237224,1036.388157,27.821927,7.534800e+02,172.32
6484487,6484487,7902379,16,A,2026-07-21 13:09:29.001408+00:00,1,-56.5413,161.4417,1,1898.89,...,1.337397,2026,7,2026,34.901284,1.194258,1036.614420,27.821765,7.554800e+02,172.32
6484488,6484488,7902379,16,A,2026-07-21 13:09:29.001408+00:00,1,-56.5413,161.4417,1,1948.73,...,1.337540,2026,7,2026,34.901352,1.173445,1036.842729,27.823261,7.574800e+02,172.32
